File: vibecheck-core/tests/fixtures/lru_cache/gpt.go
// Here's a simple LRU Cache implementation in Go.
// Let's use a doubly linked list combined with a hash map.

package lrucache

import "container/list"

// entry represents a single key-value pair stored in the cache.
// Here's the struct we'll place inside each linked list node.
type entry struct {
	key   string
	value int
}

// LruCache is the main cache structure.
// Let's keep the capacity, a map for lookups, and a list for ordering.
type LruCache struct {
	capacity int
	items    map[string]*list.Element
	order    *list.List
}

// Step 1: Initialize a new LRU Cache with the specified capacity.
// Here's the constructor function that sets everything up.
func NewLruCache(capacity int) *LruCache {
	// Make sure the capacity is at least 1.
	if capacity < 1 {
		capacity = 1
	}
	// Return a pointer to the newly created cache.
	return &LruCache{
		capacity: capacity,
		items:    make(map[string]*list.Element),
		order:    list.New(),
	}
}

// Step 2: Retrieve the value for a given key.
// Here's how we look up and reorder on access.
func (c *LruCache) Get(key string) (int, bool) {
	// Check if the key exists in the map.
	elem, found := c.items[key]
	if !found {
		// Return zero value and false to indicate a miss.
		return 0, false
	}
	// Step 2a: Move the accessed element to the front of the list.
	// Let's mark it as the most recently used item.
	c.order.MoveToFront(elem)
	// Step 2b: Return the value stored in this element.
	return elem.Value.(*entry).value, true
}

// Step 3: Insert or update a key-value pair in the cache.
// Here's how we handle insertion and eviction logic.
func (c *LruCache) Put(key string, value int) {
	// Step 3a: If the key already exists, update and move to front.
	if elem, found := c.items[key]; found {
		// Move the element to the front of the list.
		c.order.MoveToFront(elem)
		// Update the stored value.
		elem.Value.(*entry).value = value
		return
	}
	// Step 3b: Evict the least recently used item if at capacity.
	if c.order.Len() >= c.capacity {
		// Here's where we remove the back element (the LRU item).
		back := c.order.Back()
		if back != nil {
			// Remove from the linked list.
			c.order.Remove(back)
			// Remove from the hash map as well.
			evicted := back.Value.(*entry)
			delete(c.items, evicted.key)
		}
	}
	// Step 3c: Insert the new entry at the front of the list.
	// Let's create a new entry and push it to the front.
	newEntry := &entry{key: key, value: value}
	elem := c.order.PushFront(newEntry)
	// Store the list element reference in the map.
	c.items[key] = elem
}

// Step 4: Return the current number of items in the cache.
// Here's a simple getter for the cache size.
func (c *LruCache) Len() int {
	// Return the length of the internal map.
	return len(c.items)
}

// Step 5: Remove all entries from the cache.
// Let's reset both the map and the linked list.
func (c *LruCache) Clear() {
	// Re-initialize the map to an empty state.
	c.items = make(map[string]*list.Element)
	// Re-initialize the linked list.
	c.order.Init()
}

File: vibecheck-core/tests/fixtures/lru_cache/gpt.rs
// Here's a simple LRU Cache implementation in Rust.
// Let's use a HashMap combined with a VecDeque to track access order.

use std::collections::HashMap;
use std::collections::VecDeque;

// Here's the main struct that represents our LRU Cache.
pub struct LruCache<K, V> {
    capacity: usize,
    map: HashMap<K, V>,
    order: VecDeque<K>,
}

impl<K: Eq + std::hash::Hash + Clone, V> LruCache<K, V> {
    // Step 1: Initialize the cache with a given capacity.
    // Let's make sure the capacity is at least 1.
    pub fn new(capacity: usize) -> Self {
        // Here's where we construct the empty cache.
        let capacity = if capacity == 0 { 1 } else { capacity };
        LruCache {
            capacity,
            map: HashMap::new(),
            order: VecDeque::new(),
        }
    }

    // Step 2: Check if a key exists and return a reference to the value.
    // Let's also move the accessed key to the back of the queue.
    pub fn get(&mut self, key: &K) -> Option<&V> {
        // Check if the key is present in the map.
        if self.map.contains_key(key) {
            // Step 2a: Remove the key from its current position in the order.
            self.order.retain(|k| k != key);
            // Step 2b: Push it to the back to mark it as recently used.
            self.order.push_back(key.clone());
            // Return the value from the map.
            return self.map.get(key);
        }
        // If the key is not found, return None.
        None
    }

    // Step 3: Insert a key-value pair into the cache.
    // Here's how we handle eviction when the cache is full.
    pub fn put(&mut self, key: K, value: V) {
        // Step 3a: If the key already exists, update it.
        if self.map.contains_key(&key) {
            // Remove old position from the order tracking.
            self.order.retain(|k| k != &key);
        } else if self.map.len() >= self.capacity {
            // Step 3b: Evict the least recently used item.
            // Let's pop the front of the deque, that's the LRU key.
            if let Some(evicted) = self.order.pop_front() {
                // Remove the evicted entry from the map.
                self.map.remove(&evicted);
            }
        }
        // Step 3c: Insert the new key-value pair.
        self.map.insert(key.clone(), value);
        // Mark it as the most recently used.
        self.order.push_back(key);
    }

    // Step 4: Return the current number of items in the cache.
    // Here's a simple helper method for checking the size.
    pub fn len(&self) -> usize {
        // Return the length of the internal map.
        self.map.len()
    }

    // Step 5: Check if the cache is empty.
    // Let's just compare the length to zero.
    pub fn is_empty(&self) -> bool {
        // Return true if there are no items stored.
        self.map.is_empty()
    }

    // Step 6: Clear all entries from the cache.
    // Here's how we reset everything back to the initial state.
    pub fn clear(&mut self) {
        // Remove all entries from the map.
        self.map.clear();
        // Remove all tracked keys from the order deque.
        self.order.clear();
    }
}
